src/goLanguageServer.ts

```diff
@@ -6,92 +6,118 @@
-import { GoImplementationProvider } from './goImplementations';
 import { promptForMissingTool, promptForUpdatingTool } from './goInstallTools';
-import { parseLiveFile } from './goLiveErrors';
-import { GO_MODE } from './goMode';
-import { GoDocumentSymbolProvider } from './goOutline';
 import { getToolFromToolPath } from './goPath';
-import { GoReferenceProvider } from './goReferences';
-import { GoRenameProvider } from './goRename';
-import { GoSignatureHelpProvider } from './goSignature';
-import { GoCompletionItemProvider } from './goSuggest';
-import { GoWorkspaceSymbolProvider } from './goSymbol';
 import { getTool, Tool } from './goTools';
-import { GoTypeDefinitionProvider } from './goTypeDefinition';
 import { getBinPath, getCurrentGoPath, getGoConfig, getToolsEnvVars } from './util';
 
 interface LanguageServerConfig {
+	serverName: string;
+	path: string;
 	enabled: boolean;
 	flags: string[];
+	env: any;
 	features: {
 		diagnostics: boolean;
 		documentLink: boolean;
 	};
 	checkForUpdates: boolean;
 }
 
-// registerLanguageFeatures registers providers for all the language features.
-// It looks to either the language server or the standard providers for these features.
-export async function registerLanguageFeatures(ctx: vscode.ExtensionContext) {
+// Global variables used for management of the language client.
+// They are global so that the server can be easily restarted with
+// new configurations.
+let languageClient: LanguageClient;
+let languageServerDisposable: vscode.Disposable;
+let latestConfig: LanguageServerConfig;
+let serverOutputChannel: vscode.OutputChannel;
+
+// startLanguageServer starts the language server (if enabled), returning
+// true on success.
+export async function startLanguageServer(ctx: vscode.ExtensionContext): Promise<boolean> {
 	// Subscribe to notifications for changes to the configuration of the language server.
 	ctx.subscriptions.push(vscode.workspace.onDidChangeConfiguration((e) => watchLanguageServerConfiguration(e)));
 
-	const config = parseLanguageServerConfig();
+	// Support a command to restart the language server.
+	ctx.subscriptions.push(vscode.commands.registerCommand('go.languageserver.restart', () => {
+		return restartLanguageServer(ctx, parseLanguageServerConfig());
+	}));
 
-	// If the user has not enabled the language server,
-	// register the default language features and return.
-	if (!config.enabled) {
-		registerUsualProviders(ctx);
-		return;
+	const config = parseLanguageServerConfig();
+	if (!config || !config.enabled) {
+		return false;
+	}
+	const tool = getTool(config.serverName);
+	if (!tool) {
+		return false;
 	}
+	const versionToUpdate = await shouldUpdateLanguageServer(tool, config.path, config.checkForUpdates);
+	if (versionToUpdate) {
+		promptForUpdatingTool(tool.name);
+	}
+	// This function handles the case when the server isn't started yet,
+	// so we can call it to start the language server.
+	return restartLanguageServer(ctx, config);
```

Comment:
It is odd that we are using a method named "restart.." to start the language server for the first time as well.
Another oddity is that `latestConfig` is never initialized. Meaning, the first time it gets a value set to itself is when the check `if (!deepEqual(latestConfig, config))` fails. 

I get that functionally there is nothing wrong with this, as you have ensured both the first start and subsequent restarts are doing the right things.

But, for the sake of readability, I would recommend the below refactoring

- Rename `startLanguageServer` to `registerLanguageFeatures`. Initialize `latestConfig` here.
- Rename `rebuildLanguageClient` to `startLanguageServer`. Skip building the language client if config has not changed and  start language server here.
- Update `restartLanguageServer` to call `startLanguageServer`

Reply:
I made the renaming changes, but I'm not sure I agree that the starting the language client in `rebuildLanguageClient` will make things more readable. Moving the language client start into `rebuildLanguageClient` would basically collapse it into `startLanguageServer`, and I think that having a separate function for building the very large language client object makes things cleaner.

My intention is to make the code in `startLanguageServer` more maintainable - any start of the language server follows the same code path, so we will never have any special cases to fix or reason about.

Similarly, I don't think it's necessary to explicitly initialize `latestConfig` - I don't see a reason why it is different from the other global variables which aren't explicitly initialized. Moreover, it will mean that we may have to remember to fix multiple places in the code if we want to make a change.

Of course I'm happy to change things as you prefer or add more documentation to make things easier to understand, but I just wanted to clarify my thinking before doing that.

Reply:
> Moving the language client start into rebuildLanguageClient would basically collapse it into startLanguageServer, and I think that having a separate function for building the very large language client object makes things cleaner.

makes sense. The only change now I would suggest is to rename `rebuildLanguageClient` to `buildLanguageClient` since it gets used for the initial build as well

Reply:
Done. Thanks!